### PR TITLE
[IoTDB-468] Restructure QueryPlan

### PR DESCRIPTION
**Pull request by Alima777**
The AlignByDevice query ( which is called groupByDevice before ) and general query aligning by time are both storaged in QueryPlan now. However, they invoke different query process bascially, so it's necessary to restructure the QueryPlan and seperate the AlignByDeviceQuery from it.

The QueryPlan is designed as base class, which has RawDataQueryPlan and AlignByDevicePlan as subclasses. This restructure may bring some usage changes, because the general query which uses QueryPlan directly before has to instantiate RawDataQueryPlan now.

If you think the design is not friendly to users, welcome to comment.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 1]
```diff
@@ -80,39 +84,39 @@
...
+    this.measurementColumnsGroupByDevice = alignByDevicePlan.getDeviceToMeasurementsMap();
+    this.deviceToFilterMap = alignByDevicePlan.getDeviceToFilterMap();
+    this.notExistMeasurements = alignByDevicePlan.getNotExistMeasurements();
+    this.constMeasurements = alignByDevicePlan.getConstMeasurements();
+    this.positionOfNotExistMeasurements = alignByDevicePlan.getPositionOfNotExistMeasurements();
+    this.positionOfConstMeasurements = alignByDevicePlan.getPositionOfConstMeasurements();
     //BuildOutDataTypes();
 
-    if (queryPlan instanceof GroupByPlan) {
+    if (alignByDevicePlan.getGroupByPlan() != null) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: We could add a queryType in AlignByDevicePlan, and use switch here

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 2]
```diff
@@ -207,7 +211,7 @@ protected boolean hasNextWithoutConstraint() throws IOException {
             currentDataSet = queryRouter.fill(fillQueryPlan, context);
             break;
           case QUERY:
-            QueryPlan queryPlan = new QueryPlan();
+            RawDataQueryPlan queryPlan = new RawDataQueryPlan();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Don't you save three plans in AlignByDevicePlan? They could be reused

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 3]
```diff
@@ -80,39 +84,39 @@
...
 
-    if (queryPlan instanceof GroupByPlan) {
+    if (alignByDevicePlan.getGroupByPlan() != null) {
       this.dataSetType = DataSetType.GROUPBY;
       // assign parameters
-      this.unit = ((GroupByPlan) queryPlan).getInterval();
-      this.slidingStep = ((GroupByPlan) queryPlan).getSlidingStep();
-      this.startTime = ((GroupByPlan) queryPlan).getStartTime();
-      this.endTime = ((GroupByPlan) queryPlan).getEndTime();
+      this.unit = alignByDevicePlan.getGroupByPlan().getInterval();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Reuse the query plan in AlignByDevicePlan, no need to set the parameters again and again

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import java.util.Map;
+import java.util.Set;
+import org.apache.iotdb.db.qp.logical.Operator;
+import org.apache.iotdb.db.qp.logical.Operator.OperatorType;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.expression.IExpression;
+
+public class AlignByDevicePlan extends QueryPlan {
+
+  private List<String> measurements; // for group by device sql, e.g. temperature
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Actually, the paths in QueryPlan are just for verification now and the acutal paths and datatypes are calculated for every device in DataSet. I will try reorganize the structure.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  private Map<String, Set<String>> deviceToMeasurementsMap; // for group by device sql, e.g. root.ln.d1 -> temperature
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  private Map<String, TSDataType> dataTypeConsistencyChecker; // for group by device sql, e.g. root.ln.d1.temperature -> Float
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Add a comment to explain it.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  private Map<String, IExpression> deviceToFilterMap; // for group by device sql
+
+  private GroupByPlan groupByPlan;
+  private FillQueryPlan fillQueryPlan;
+  private AggregationPlan aggregationPlan;
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Of course there isn't a RawDataQueryPan. We don't need parameters in it.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+  //the measurements that do not exist in any device,
+  // data type is considered as Boolean. The value is considered as null
+  private List<String> notExistMeasurements = new ArrayList<>();
+  ; // for group by device sql
+  private List<Integer> positionOfNotExistMeasurements = new ArrayList<>(); // for group by device sql
+  //the measurements that have quotation mark. e.g., "abc",
+  // '11', the data type is considered as String and the value  is considered is the same with measurement name
+  private List<String> constMeasurements = new ArrayList<>(); // for group by device sql
+  private List<Integer> positionOfConstMeasurements = new ArrayList<>(); // for group by device sql
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: OK.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+  public void setAggregationPlan(AggregationPlan aggregationPlan) {
+    this.aggregationPlan = aggregationPlan;
+    this.setOperatorType(Operator.OperatorType.AGGREGATION);
+  }
+
+  //the measurements that do not exist in any device,
+  // data type is considered as Boolean. The value is considered as null
+  private List<String> notExistMeasurements = new ArrayList<>();
+  ; // for group by device sql
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  //suppose user writes SELECT 'c1',a1,b1,b2,'c2',a2,a3,'c3',b3,a4,a5 FROM ... where for each a_i
+  // column there is at least one device having it, and for each b_i column there is no device
+  // having it, and 'c_i' is a const column.
+  // Then, measurements = {a1, a2, a3, a4, a5};
+  // notExistMeasurements = {b1, b2, b3}, and positionOfNotExistMeasurements = {2, 3, 8};
+  // constMeasurements = {'c1', 'c2', 'c3'}, and positionOfConstMeasurements = {0, 4, 7}.
+  // When to reproduce the order of measurements. The pseudocode is:
+  //<pre>
+  // current = 0;
+  // if (min(notExist, const) <= current) {
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: I don't understand it too. Let's call that guy writing this comment to modify.

File: server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java [hunk 11]
```diff
@@ -242,16 +244,26 @@ private PhysicalPlan transformQuery(QueryOperator queryOperator)
       ((AggregationPlan) queryPlan)
           .setAggregations(queryOperator.getSelectOperator().getAggregations());
     } else {
-      queryPlan = new QueryPlan();
+      queryPlan = new RawDataQueryPlan();
     }
     if (queryOperator.isGroupByDevice()) {
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java]: I think it's nice now.

File: server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java [hunk 12]
```diff
@@ -373,17 +386,17 @@ private PhysicalPlan transformQuery(QueryOperator queryOperator)
       }
 
       // assigns to queryPlan
-      queryPlan.setGroupByDevice(true);
-      queryPlan.setMeasurements(measurements);
-      queryPlan.setMeasurementsGroupByDevice(measurementsGroupByDevice);
-      queryPlan.setDataTypeConsistencyChecker(dataTypeConsistencyChecker);
+      ((AlignByDevicePlan) queryPlan).setMeasurements(measurements);
+      ((AlignByDevicePlan) queryPlan).setMeasurementsGroupByDevice(deviceToMeasurementsMap);
+      ((AlignByDevicePlan) queryPlan).setDataTypeConsistencyChecker(dataTypeConsistencyChecker);
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java [hunk 13]
```diff
@@ -242,16 +244,26 @@ private PhysicalPlan transformQuery(QueryOperator queryOperator)
...
+      AlignByDevicePlan alignByDevicePlan = new AlignByDevicePlan();
+      if (queryPlan instanceof GroupByPlan) {
+        alignByDevicePlan.setGroupByPlan((GroupByPlan) queryPlan);
+      } else if (queryPlan instanceof FillQueryPlan) {
+        alignByDevicePlan.setFillQueryPlan((FillQueryPlan) queryPlan);
+      } else if (queryPlan instanceof AggregationPlan) {
+        alignByDevicePlan.setAggregationPlan((AggregationPlan) queryPlan);
+      }
+
+      queryPlan = alignByDevicePlan;
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java]: Thank you for advice! Fixed.

File: server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java [hunk 14]
```diff
@@ -395,7 +408,7 @@ private PhysicalPlan transformQuery(QueryOperator queryOperator)
 
       if (filterOperator != null) {
         IExpression expression = filterOperator.transformToExpression();
-        queryPlan.setExpression(expression);
+        ((RawDataQueryPlan) queryPlan).setExpression(expression);
       }
     }
     generateDataTypes(queryPlan);
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/strategy/PhysicalGenerator.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 15]
```diff
@@ -80,39 +84,39 @@
   private int[] currentColumnMapRelation;
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 16]
```diff
@@ -80,39 +84,39 @@
...
+      this.endTime = alignByDevicePlan.getGroupByPlan().getEndTime();
 
-    } else if (queryPlan instanceof AggregationPlan) {
+    } else if (alignByDevicePlan.getAggregationPlan() != null) {
       this.dataSetType = DataSetType.AGGREGATE;
 
-    } else if (queryPlan instanceof FillQueryPlan) {
+    } else if (alignByDevicePlan.getFillQueryPlan() != null) {
       this.dataSetType = DataSetType.FILL;
       // assign parameters
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java (hunk 1 above)
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java (hunk 3 above)
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Good Idea. Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java (hunk 2 above)
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Good Idea. Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 17]
```diff
@@ -80,39 +84,39 @@
...
-    this.constMeasurements = queryPlan.getConstMeasurements();
-    this.positionOfNotExistMeasurements = queryPlan.getPositionOfNotExistMeasurements();
-    this.positionOfConstMeasurements = queryPlan.getPositionOfConstMeasurements();
+    this.measurementColumnsGroupByDevice = alignByDevicePlan.getDeviceToMeasurementsMap();
+    this.deviceToFilterMap = alignByDevicePlan.getDeviceToFilterMap();
+    this.notExistMeasurements = alignByDevicePlan.getNotExistMeasurements();
+    this.constMeasurements = alignByDevicePlan.getConstMeasurements();
+    this.positionOfNotExistMeasurements = alignByDevicePlan.getPositionOfNotExistMeasurements();
+    this.positionOfConstMeasurements = alignByDevicePlan.getPositionOfConstMeasurements();
     //BuildOutDataTypes();
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 18]
```diff
@@ -80,39 +84,39 @@
...
 
-    if (queryPlan instanceof GroupByPlan) {
+    if (alignByDevicePlan.getGroupByPlan() != null) {
       this.dataSetType = DataSetType.GROUPBY;
       // assign parameters
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 19]
```diff
@@ -36,9 +42,6 @@
 import org.apache.iotdb.tsfile.read.query.dataset.QueryDataSet;
 import org.apache.iotdb.tsfile.utils.Binary;
 
-import java.io.IOException;
-import java.util.*;
-
 
 /**
  * This QueryDataSet is used for GROUP_BY_DEVICE query result.
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]:
```suggestion
 * This QueryDataSet is used for align by device query result.
```

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java [hunk 20]
```diff
@@ -80,45 +77,41 @@
   private int[] currentColumnMapRelation;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: how about renaming this class to AlignByDeviceDataset

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jixuan1989) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Hi, I am the author :-D 

Notice that:
```
positionOfNotExistMeasurements = {2, 3, 8} 
positionOfConstMeasurements = {0, 4, 7}
```
Then
```
if (min(notExist, const) <= current)
```
means `if min(positionOfNotExistMeasurements[i], positionOfConstMeasurements[j]) < current`

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+  // the measurements that have quotation mark. e.g. "abc",
+  // '11', the data type is considered as String and the value is considered is the same with measurement name
+  private List<String> constMeasurements = new ArrayList<>();
+  private List<Integer> positionOfConstMeasurements = new ArrayList<>();
+
+  public AlignByDevicePlan() {
+    super();
+  }
+
+  public AlignByDevicePlan(boolean isQuery, Operator.OperatorType operatorType) {
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/GroupByPlan.java [hunk 22]
```diff
@@ -20,7 +20,8 @@
 
 import org.apache.iotdb.db.qp.logical.Operator;
 
-public class GroupByPlan extends AggregationPlan {
+public class
+GroupByPlan extends AggregationPlan {
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/GroupByPlan.java]: ...Fixed.

[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/GroupByPlan.java]: Just keep it.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java (hunk 19 above)
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java (hunk 20 above)
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java]: It's nice. I will rename it.

File: server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java [hunk 23]
```diff
@@ -45,8 +45,8 @@
   private int seriesLimit = 0;
   private int seriesOffset = 0;
 
-  private boolean isGroupByDevice = false;
-  private boolean isAlign = true;
+  private boolean isAlignByDevice = false;
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java]: Just keep it.

File: server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java [hunk 24]
```diff
@@ -45,8 +45,8 @@
...
+  private boolean isAlignByTime = true;
```
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java]: Just keep it please.

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Alima777) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/AlignByDevicePlan.java]: I have added a RawDataQueryPlan in AlignByDeviceDataSet.

File: server/src/main/java/org/apache/iotdb/db/query/dataset/AlignByDeviceDataSet.java [hunk 25]
```diff
@@ -80,45 +78,42 @@
...
+        this.dataSetType = DataSetType.AGGREGATE;
+        this.aggregationPlan = alignByDevicePlan.getAggregationPlan();
+        break;
+      case FILL:
+        this.dataSetType = DataSetType.FILL;
+        this.fillQueryPlan = alignByDevicePlan.getFillQueryPlan();
+        break;
+      default:
+        this.dataSetType = DataSetType.QUERY;
+        this.rawDataQueryPlan = new RawDataQueryPlan();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/AlignByDeviceDataSet.java]: This logic is a little strange

**Pull request merged**